File: day07/CPP/day07.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>
#include <cstdio>
#include <bits/stdc++.h>
#include <cstring>
#include <cstdlib>
#include <unistd.h>
#include <fstream>

using namespace std;
void print_title() {
    string stuff(25, '-');
    cout << "\033[1;33m";
    cout << stuff <<"Advent of Code - Day 7"  << stuff  << endl;
    cout << "\033[0m";
    cout << string(40, ' ') << endl;
    // run terminal command in c++
    string cmd = "figlet The Treachery of Whales -c -f small";
    char *command_char = new char[cmd.length() + 1];
    strcpy(command_char, cmd.c_str());
    // store the output of the command in a string
    string output = "";
    char buffer[128];
    FILE *fp;
    fp = popen(command_char, "r");
    if (fp == NULL)
    {
        cout << "Failed to run command" << endl;
    }
    while (fgets(buffer, sizeof(buffer), fp) != NULL)
    {
        output += buffer;
    }
    pclose(fp);
    cout << "\033[0;32m";
    cout << output << endl;
    cout << "\033[0m" << endl;
    string stufff(33, '-');
    cout << "\033[1;33m";
    cout << stufff <<"Output"  << stufff  << endl;
    cout << "\033[0m" << endl;
}
int main() {
    print_title();
    vector<int> pos;
    ifstream f("../Input/day7.txt");
    while (f) {
        int p;
        f >> p;
        if (!f) break;
        pos.push_back(p); 
        f.ignore(1);
    }

    auto minmax_Pos = minmax_element(pos.cbegin(), pos.cend());

    int fuel_min = 1000000000;
    for (int i=*minmax_Pos.first;i<=*minmax_Pos.second;i++) {
        int fuel = 0;
        for (auto p : pos) {
            fuel += abs(p-i);
        }
        fuel_min = min(fuel, fuel_min);
    }
    int fuel_min_2 = 1000000000;
    for (int i=*minmax_Pos.first;i<=*minmax_Pos.second;i++) {
        int fuel = 0;
        for (auto p : pos) {
            int dist = abs(p-i);
            fuel += (dist*(1+dist))/2;
        }
        fuel_min_2 = min(fuel, fuel_min_2);
    }

    cout << "Puzzle 1: " << fuel_min << endl;
    cout << "Puzzle 2: " << fuel_min_2 << endl;
    string stuff_final(72, '=');
    cout << endl;
    cout << stuff_final << endl;

}

File: day07/Kotlin/day07.kt
import java.io.*
import kotlin.math.abs
fun print_day_7() {
    
    val yellow = "\u001B[33m"
    val reset = "\u001b[0m"
    val green = "\u001B[32m"
    println(yellow + "-".repeat(25) + "Advent of Code - Day 7" + "-".repeat(25) + reset)
    println(green)
    val process = Runtime.getRuntime().exec("figlet The Treachery of Whales -c -f small")
    val reader = BufferedReader(InputStreamReader(process.inputStream))
    reader.forEachLine { println(it) }
    println(reset)

    println(yellow + "-".repeat(33) + "Output" + "-".repeat(33) + reset + "\n")
}
fun main() {
    print_day_7()
    Day7().part_1()
    Day7().part_2()
    println("\n" + "\u001B[33m" + "=".repeat(72) + "\u001b[0m" + "\n")
}

class Day7 {
    val crabs = File("../Input/day7.txt").readLines().first().split(",").map{it.toInt()}

    fun part_1() {
        val min = crabs.minOrNull()!!
        val max = crabs.maxOrNull()!!

        var lowest = Int.MAX_VALUE

        for(i in min .. max) {
            var fuelUsed = 0
            for(crab in crabs) { fuelUsed += abs(crab - i) }

            if (fuelUsed < lowest) {
                lowest = fuelUsed
            }
        }
        println("Puzzle 1: $lowest")
    }

    fun part_2() {

        val min = crabs.minOrNull()!!
        val max = crabs.maxOrNull()!!

        var lowest = Int.MAX_VALUE

        for(i in min .. max) {
            var fuelUsed = 0
            for(crab in crabs) { fuelUsed += calcFibIncrement(abs(crab - i)) }

            if (fuelUsed < lowest) {
                lowest = fuelUsed
            }
        }
        println("Puzzle 2: $lowest")
    }
    fun calcFibIncrement(increment: Int, incrementBy: Int = 0): Int {
        return if (increment == 0)
            incrementBy
        else
            calcFibIncrement(increment - 1, incrementBy + 1) + incrementBy
    }

}